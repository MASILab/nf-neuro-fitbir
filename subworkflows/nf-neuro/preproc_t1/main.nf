// ** Importing modules from nf-neuro ** //
include { DENOISING_NLMEANS } from '../../../modules/nf-neuro/denoising/nlmeans/main'
include { PREPROC_N4 } from '../../../modules/nf-neuro/preproc/n4/main'
include { IMAGE_RESAMPLE } from '../../../modules/nf-neuro/image/resample/main'
include { BETCROP_ANTSBET } from '../../../modules/nf-neuro/betcrop/antsbet/main'
include { BETCROP_SYNTHBET} from '../../../modules/nf-neuro/betcrop/synthbet/main'
include { IMAGE_CROPVOLUME as IMAGE_CROPVOLUME_T1 } from '../../../modules/nf-neuro/image/cropvolume/main'
include { IMAGE_CROPVOLUME as IMAGE_CROPVOLUME_MASK } from '../../../modules/nf-neuro/image/cropvolume/main'

params.preproc_t1_run_synthbet = false

workflow PREPROC_T1 {

    take:
        ch_image            // channel: [ val(meta), image ]
        ch_template         // channel: [ val(meta), template ]         , optional
        ch_probability_map  // channel: [ val(meta), probability-map ]  , optional
        ch_template_mask    // channel: [ val(meta), mask ]             , optional
        ch_initial_affine   // channel: [ val(meta), init_affine ]      , optional
        ch_ref_mask         // channel: [ val(meta), mask ]             , optional
        ch_ref_resample     // channel: [ val(meta), ref ]              , optional
        ch_weights          // channel: [ val(meta), weights ]          , optional
    main:
        ch_versions = Channel.empty()
        image_nlmeans = Channel.empty()
        image_N4 = Channel.empty()
        image_resample = Channel.empty()
        image_bet = Channel.empty()
        mask_bet = Channel.empty()
        image_crop = Channel.empty()
        mask_crop = Channel.empty()
        bbox = Channel.empty()
        ch_mask = Channel.empty()

        if ( params.preproc_t1_run_denoising ) {

            ch_nlmeans = ch_image
                .join(ch_ref_mask, remainder: true)
                .map{ meta, image, mask -> [meta, image, mask ?: [], []] }

            DENOISING_NLMEANS ( ch_nlmeans )
            ch_versions = ch_versions.mix(DENOISING_NLMEANS.out.versions.first())
            image_nlmeans = DENOISING_NLMEANS.out.image
            ch_image = DENOISING_NLMEANS.out.image
        }

        if ( params.preproc_t1_run_N4 ) {
            ch_N4 = ch_image
                .map{ meta, image -> [meta, image, [], []] }
                .join(ch_ref_mask, remainder: true)
                .map{ meta, image, bval, bvec, mask -> [meta, image, bval, bvec, mask ?: []] }

            PREPROC_N4 ( ch_N4 )
            ch_versions = ch_versions.mix(PREPROC_N4.out.versions.first())
            image_N4 = PREPROC_N4.out.image
            ch_image = PREPROC_N4.out.image
        }

        if ( params.preproc_t1_run_resampling ) {
            ch_resampling = ch_image
                .join(ch_ref_resample, remainder: true)
                .map{ meta, image, ref_image -> [meta, image, ref_image ?: []] }

            IMAGE_RESAMPLE ( ch_resampling )
            ch_versions = ch_versions.mix(IMAGE_RESAMPLE.out.versions.first())
            image_resample = IMAGE_RESAMPLE.out.image
            ch_image = IMAGE_RESAMPLE.out.image
        }

        if ( params.preproc_t1_run_synthbet ) {
            ch_bet = ch_image
                .join(ch_weights, remainder: true)
                .map{ meta, image, weights -> [meta, image, weights ?: []] }

            BETCROP_SYNTHBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_SYNTHBET.out.versions.first())

            // ** Setting BET output ** //
            image_bet = BETCROP_SYNTHBET.out.bet_image
            ch_image = BETCROP_SYNTHBET.out.bet_image
            mask_bet = BETCROP_SYNTHBET.out.brain_mask
            ch_mask = BETCROP_SYNTHBET.out.brain_mask
        }
        else if ( params.preproc_t1_run_ants_bet ) {
            // ** ANTSBET ** //
            // The template and probability maps are mandatory if running antsBET. Since the
            // error message from nextflow when they are absent is either non-informative or
            // missing, we use ifEmpty to provide a more informative one.
            ch_bet = ch_image
                .join(ch_template.ifEmpty{ error("ANTS BET needs a template") })
                .join(ch_probability_map.ifEmpty{ error("ANTS BET needs a tissue probability map") })
                .join(ch_template_mask, remainder: true)
                .map{ meta, image, template, probability_map, mask -> [meta, image, template, probability_map, mask ?: []] }
                .join(ch_initial_affine, remainder: true)
                .map{ meta, image, template, probability_map, mask, init_affine -> [meta, image, template, probability_map, mask, init_affine ?: []] }

            BETCROP_ANTSBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_ANTSBET.out.versions.first())

            // ** Setting BET output ** //
            image_bet = BETCROP_ANTSBET.out.t1
            ch_image = BETCROP_ANTSBET.out.t1
            mask_bet = BETCROP_ANTSBET.out.mask
            ch_mask = BETCROP_ANTSBET.out.mask
        }

        if ( params.preproc_t1_run_crop ) {
            // ** Crop image ** //
            ch_crop = ch_image
                .map{ it + [[]] }

            IMAGE_CROPVOLUME_T1 ( ch_crop )
            ch_versions = ch_versions.mix(IMAGE_CROPVOLUME_T1.out.versions.first())
            image_crop = IMAGE_CROPVOLUME_T1.out.image
            ch_image = IMAGE_CROPVOLUME_T1.out.image
            bbox = IMAGE_CROPVOLUME_T1.out.bounding_box

            // ** Crop mask ** //
            ch_crop_mask = mask_bet
                .join(IMAGE_CROPVOLUME_T1.out.bounding_box)

            IMAGE_CROPVOLUME_MASK ( ch_crop_mask )
            ch_versions = ch_versions.mix(IMAGE_CROPVOLUME_MASK.out.versions.first())
            mask_crop = IMAGE_CROPVOLUME_MASK.out.image
            ch_mask = IMAGE_CROPVOLUME_MASK.out.image
        }

    emit:
        t1_final        = ch_image          // channel: [ val(meta), t1-preprocessed ]
        mask_final      = ch_mask           // channel: [ val(meta), t1-mask ]
        image_nlmeans   = image_nlmeans     // channel: [ val(meta), t1-after-denoise ]
        image_N4        = image_N4          // channel: [ val(meta), t1-after-unbias ]
        image_resample  = image_resample    // channel: [ val(meta), t1-after-resample ]
        image_bet       = image_bet         // channel: [ val(meta), t1-after-bet ]
        mask_bet        = mask_bet          // channel: [ val(meta), intermediary-mask ]
        crop_box        = bbox              // channel: [ val(meta), bounding-box ]
        versions        = ch_versions       // channel: [ versions.yml ]
}
