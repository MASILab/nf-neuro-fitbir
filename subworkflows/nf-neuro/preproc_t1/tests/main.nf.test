nextflow_workflow {

    name "Test Subworkflow PREPROC_T1"
    script "../main.nf"
    workflow "PREPROC_T1"

    tag "subworkflows"
    tag "subworkflows_nfneuro"
    tag "subworkflows/preproc_t1"

    tag "denoising/nlmeans"
    tag "preproc/n4"
    tag "image/resample"
    tag "betcrop/antsbet"
    tag "betcrop/synthbet"
    tag "image/cropvolume"

    tag "load_test_data"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "antsbet.zip" , "T1w.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("preproc_t1_antsbet_error") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assert workflow.failed
        }
    }

    test("preproc_t1_antsbet") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[2] = ch_split_test_data.antsbet.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz"),
                    ]}
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel ->
                        channel.value.every{ subject -> subject instanceof ArrayList
                            ? subject.every()
                            : subject } } }
            )
        }
    }

    test("preproc_t1_synthbet") {
        config "./nextflow_synthbet.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel ->
                        channel.value.every{ subject -> subject instanceof ArrayList
                            ? subject.every()
                            : subject } } }
            )
        }
    }

    test("preproc_t1_quick") {
        config "./nextflow_quick.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["image_N4",
                                        "image_bet",
                                        "mask_bet"].contains(channel.key)
                        ? channel.value.size() == 0
                        : channel.value.every{ subject -> subject instanceof ArrayList
                            ? subject.every()
                            : subject } } }
            )
        }
    }

    test("preproc_t1_skip_all") {
        config "./nextflow_skip_all.config"

        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        antsbet: it.simpleName == "antsbet"
                        t1w: it.simpleName == "T1w"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["image_nlmeans",
                                        "image_N4",
                                        "image_resample",
                                        "image_bet",
                                        "mask_bet",
                                        "crop_box"].contains(channel.key)
                        ? channel.value.size() == 0
                        : channel.value.every{ subject -> subject instanceof ArrayList
                            ? subject.every()
                            : subject } } }
            )
        }
    }
}
