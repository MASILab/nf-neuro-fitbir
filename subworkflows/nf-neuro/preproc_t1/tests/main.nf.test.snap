{
    "preproc_t1_quick": {
        "content": [
            {
                "image_nlmeans": [
                    [
                        {
                            "id": "test"
                        },
                        "test__denoised.nii.gz"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test_resampled.nii.gz"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9d7a3e397fd20bc0d4b8b46d62060806",
                    "versions.yml:md5,b539f60cb48d0e92cd7dcf2f403d3ea9",
                    "versions.yml:md5,f2536b7124723d459cfbbb79fcd944cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-24T18:51:25.704452222"
    },
    "preproc_t1_synthbet": {
        "content": [
            {
                "crop_box": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped_bbox.pkl"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__image_n4.nii.gz"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__bet_image.nii.gz"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test"
                        },
                        "test__denoised.nii.gz"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test_resampled.nii.gz"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__brain_mask.nii.gz"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cropped.nii.gz"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,318cabe934be45528a25f52083d9c90d",
                    "versions.yml:md5,7aad91b5043d2dc8e8fc69ffbbe6f657",
                    "versions.yml:md5,9d7a3e397fd20bc0d4b8b46d62060806",
                    "versions.yml:md5,b539f60cb48d0e92cd7dcf2f403d3ea9",
                    "versions.yml:md5,edd3c88d7ff0ed68c873953116ed3cb0",
                    "versions.yml:md5,f2536b7124723d459cfbbb79fcd944cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-24T18:51:05.520508565"
    },
    "preproc_t1_skip_all": {
        "content": [
            {
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "T1w.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-27T12:18:38.832987377"
    },
    "preproc_t1_antsbet": {
        "content": [
            {
                "crop_box": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped_bbox.pkl"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__image_n4.nii.gz"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_bet.nii.gz"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test"
                        },
                        "test__denoised.nii.gz"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test_resampled.nii.gz"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_bet_mask.nii.gz"
                    ]
                ],
                "mask_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_cropped.nii.gz"
                    ]
                ],
                "t1_final": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3bda0d9faae5e07d500a04a0119b766f",
                    "versions.yml:md5,7aad91b5043d2dc8e8fc69ffbbe6f657",
                    "versions.yml:md5,9d7a3e397fd20bc0d4b8b46d62060806",
                    "versions.yml:md5,b539f60cb48d0e92cd7dcf2f403d3ea9",
                    "versions.yml:md5,edd3c88d7ff0ed68c873953116ed3cb0",
                    "versions.yml:md5,f2536b7124723d459cfbbb79fcd944cc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-24T18:50:33.676292448"
    }
}